### PR TITLE
UCT/RC/DC: Support SRQ linked list reordering

### PR DESCRIPTION
**Pull request by Artemy-Mellanox**

### CODE REVIEW

File: src/uct/ib/rc/accel/rc_mlx5_common.c [hunk 1]
```diff
@@ -107,35 +153,48 @@ unsigned uct_rc_mlx5_iface_srq_post_recv(uct_rc_mlx5_iface_common_t *iface)
...
+}
+
+unsigned uct_rc_mlx5_iface_srq_post_recv_linked_list(
+        uct_rc_mlx5_iface_common_t *iface)
+{
+    uct_ib_mlx5_srq_t *srq   = &iface->rx.srq;
+    uct_rc_iface_t *rc_iface = &iface->super;
+    uct_ib_mlx5_srq_seg_t *seg = NULL;
+    uint16_t count = 0;
+    uint16_t wqe_index, next_index;
```
Comment (yosefe) [inline src/uct/ib/rc/accel/rc_mlx5_common.c:175]: indent =

File: src/uct/ib/rc/accel/rc_mlx5_common.c [hunk 2]
```diff
@@ -107,35 +153,48 @@ unsigned uct_rc_mlx5_iface_srq_post_recv(uct_rc_mlx5_iface_common_t *iface)
...
+    ucs_assert(rc_iface->rx.srq.available > 0);
+
+    wqe_index = srq->ready_idx;
+    seg = uct_ib_mlx5_srq_get_wqe(srq, wqe_index);
+
+    for (;;) {
+        next_index = ntohs(seg->srq.next_wqe_index);
+        if (next_index == (srq->free_idx & srq->mask)) {
+            break;
+        }
```
Comment (yosefe) [inline src/uct/ib/rc/accel/rc_mlx5_common.c:186]: make check available instead, and make this an assertion?

File: src/uct/ib/rc/accel/rc_mlx5_common.c [hunk 3]
```diff
@@ -107,35 +153,48 @@ unsigned uct_rc_mlx5_iface_srq_post_recv(uct_rc_mlx5_iface_common_t *iface)
...
+        seg = uct_ib_mlx5_srq_get_wqe(srq, next_index);
+
+        if (uct_rc_mlx5_iface_srq_set_seg(iface, seg) != UCS_OK) {
+            break;
+        }
+
+        wqe_index = next_index;
```
Comment (yosefe) [inline src/uct/ib/rc/accel/rc_mlx5_common.c:193]: move it outside the loop?

File: src/uct/ib/rc/accel/rc_mlx5_iface.c [hunk 4]
```diff
@@ -144,6 +146,20 @@ unsigned uct_rc_mlx5_iface_progress(void *arg)
...
+{
+    uct_rc_mlx5_iface_common_t *iface = arg;
+    unsigned count;
+
+    count = uct_rc_mlx5_iface_common_poll_rx(iface,
+                                             UCT_RC_MLX5_POLL_FLAG_HAS_EP |
+                                             UCT_RC_MLX5_POLL_FLAG_LINKED_LIST);
+    if (!uct_rc_iface_poll_tx(&iface->super, count)) {
+        return count;
+    }
```
Comment (yosefe) [inline src/uct/ib/rc/accel/rc_mlx5_iface.c:159]: space line after

File: src/uct/ib/rc/accel/rc_mlx5_iface.c [hunk 5]
```diff
@@ -502,7 +519,11 @@ uct_rc_mlx5_iface_init_rx(uct_rc_iface_t *rc_iface,
         return status;
     }
 
-    iface->super.progress = uct_rc_mlx5_iface_progress;
+    if (iface->config.linked_list_srq) {
+        iface->super.progress = uct_rc_mlx5_iface_progress_linked_list;
+    } else {
+        iface->super.progress = uct_rc_mlx5_iface_progress;
```
Comment (yosefe) [inline src/uct/ib/rc/accel/rc_mlx5_iface.c:525]: maybe:
uct_rc_mlx5_iface_progress_cyclic_rq
uct_rc_mlx5_iface_progress_ll_rq

and similar naming for post_recv funcs

**Review by yosefe — COMMENTED** [review]
pls add some gtests which have >1 sender and >mtu message size, to run with linked list rq

File: src/uct/ib/dc/dc_mlx5.c [hunk 6]
```diff
@@ -249,6 +249,19 @@ static unsigned uct_dc_mlx5_iface_progress(void *arg)
...
+{
+    uct_dc_mlx5_iface_t *iface = arg;
+    unsigned count;
+
+    count = uct_rc_mlx5_iface_common_poll_rx(&iface->super,
+                                             UCT_RC_MLX5_POLL_FLAG_LINKED_LIST);
+    if (!uct_rc_iface_poll_tx(&iface->super.super, count)) {
+        return count;
+    }
+    return uct_dc_mlx5_poll_tx(iface);
```
Comment (dmitrygx) [inline src/uct/ib/dc/dc_mlx5.c:262]: [q] should it be sum of RX and TX poll counts?
```
return count + uct_dc_mlx5_poll_tx(iface);
```

File: src/uct/ib/rc/accel/rc_mlx5_common.c [hunk 7]
```diff
@@ -69,20 +69,66 @@ ucs_config_field_t uct_rc_mlx5_common_config_table[] = {
    ucs_offsetof(uct_rc_mlx5_iface_common_config_t, cyclic_srq_enable),
    UCS_CONFIG_TYPE_TERNARY},
 
+  {"LINKED_LIST", "n",
+   "TBD",
```
Comment (dmitrygx) [inline src/uct/ib/rc/accel/rc_mlx5_common.c:73]: add description?

File: src/uct/ib/rc/accel/rc_mlx5_common.c [hunk 8]
```diff
@@ -69,20 +69,66 @@ ucs_config_field_t uct_rc_mlx5_common_config_table[] = {
...
+{
+    uct_ib_iface_recv_desc_t *desc;
+    uint64_t desc_map;
+    void *hdr;
+    int i;
+
+    desc_map = ~seg->srq.ptr_mask & UCS_MASK(iface->tm.mp.num_strides);
+    ucs_for_each_bit(i, desc_map) {
+        UCT_TL_IFACE_GET_RX_DESC(&iface->super.super.super, &iface->super.rx.mp,
+                desc, return UCS_ERR_NO_MEMORY);
```
Comment (dmitrygx) [inline src/uct/ib/rc/accel/rc_mlx5_common.c:93]: wrong alignment?
should it be smth like?
```
ucs_for_each_bit(i, desc_map) {
    UCT_TL_IFACE_GET_RX_DESC(&iface->super.super.super, &iface->super.rx.mp,
                             desc, return UCS_ERR_NO_MEMORY);
...
```

File: src/uct/ib/rc/accel/rc_mlx5_common.c [hunk 9]
```diff
@@ -107,35 +153,48 @@ unsigned uct_rc_mlx5_iface_srq_post_recv(uct_rc_mlx5_iface_common_t *iface)
...
-        srq->sw_pi                  = wqe_index;
-        rc_iface->rx.srq.available -= count;
-        ucs_memory_cpu_store_fence();
-        *srq->db                    = htonl(srq->sw_pi);
-        ucs_assert(uct_ib_mlx5_srq_get_wqe(srq, srq->mask)->srq.next_wqe_index == 0);
+    uct_rc_mlx5_iface_update_srq_res(rc_iface, srq, wqe_index, count);
+    return count;
+}
+
+unsigned uct_rc_mlx5_iface_srq_post_recv_linked_list(
```
Comment (dmitrygx) [inline src/uct/ib/rc/accel/rc_mlx5_common.c:168]: maybe align it like this?
```
unsigned
uct_rc_mlx5_iface_srq_post_recv_linked_list(uct_rc_mlx5_iface_common_t *iface)
```

File: src/uct/ib/rc/accel/rc_mlx5_common.h [hunk 10]
```diff
@@ -583,6 +586,8 @@ uct_rc_mlx5_handle_rndv_fin(uct_rc_mlx5_iface_common_t *iface, uint32_t app_ctx)
 extern ucs_config_field_t uct_rc_mlx5_common_config_table[];
 
 unsigned uct_rc_mlx5_iface_srq_post_recv(uct_rc_mlx5_iface_common_t *iface);
+unsigned uct_rc_mlx5_iface_srq_post_recv_linked_list(
```
Comment (dmitrygx) [inline src/uct/ib/rc/accel/rc_mlx5_common.h:589]: fix alignemnt?

File: src/uct/ib/rc/accel/rc_mlx5.inl [hunk 11]
```diff
@@ -157,8 +157,13 @@ uct_rc_mlx5_iface_release_srq_seg(uct_rc_mlx5_iface_common_t *iface,
 
     seg_free = (seg->srq.ptr_mask == UCS_MASK(iface->tm.mp.num_strides));
 
-    if (ucs_likely(seg_free && (wqe_index == ((iface->rx.srq.ready_idx + 1) &
-                                              iface->rx.srq.mask)))) {
+    if (poll_flags & UCT_RC_MLX5_POLL_FLAG_LINKED_LIST) {
+        seg = uct_ib_mlx5_srq_get_wqe(&iface->rx.srq, iface->rx.srq.free_idx);
+        seg->srq.next_wqe_index = htons(wqe_index);
```
Comment (yosefe) [inline src/uct/ib/rc/accel/rc_mlx5.inl:162]: indent on column
add assert that free==0

File: src/uct/ib/rc/accel/rc_mlx5.inl [hunk 12]
```diff
@@ -155,33 +156,35 @@ uct_rc_mlx5_iface_release_srq_seg(uct_rc_mlx5_iface_common_t *iface,
...
-        seg = uct_ib_mlx5_srq_get_wqe(&iface->rx.srq, iface->rx.srq.free_idx);
+        seg                     = uct_ib_mlx5_srq_get_wqe(srq, srq->free_idx);
         seg->srq.next_wqe_index = htons(wqe_index);
-        iface->rx.srq.free_idx = wqe_index;
-    } else if (ucs_likely(seg_free &&
-                          (wqe_index == ((iface->rx.srq.ready_idx + 1) &
-                                          iface->rx.srq.mask)))) {
+        srq->free_idx           = wqe_index;
+        return;
+    }
```
Comment (yosefe) [inline src/uct/ib/rc/accel/rc_mlx5.inl:166]: nice

File: src/uct/ib/rc/accel/rc_mlx5.inl [hunk 13]
```diff
@@ -155,33 +156,35 @@ uct_rc_mlx5_iface_release_srq_seg(uct_rc_mlx5_iface_common_t *iface,
...
+
+    seg_free = (seg->srq.ptr_mask == UCS_MASK(iface->tm.mp.num_strides));
+
+    if (ucs_likely(seg_free && (wqe_ctr == srq->ready_idx + 1))) {
```
Comment (yosefe) [inline src/uct/ib/rc/accel/rc_mlx5.inl:170]: pls add ( ) around srq->ready_idx + 1

File: src/uct/ib/rc/accel/rc_mlx5.inl [hunk 14]
```diff
@@ -155,33 +156,35 @@ uct_rc_mlx5_iface_release_srq_seg(uct_rc_mlx5_iface_common_t *iface,
...
-             seg->srq.free = 1;
-         }
+         ucs_assert(wqe_ctr == srq->free_idx + 1);
+         ++srq->ready_idx;
+         ++srq->free_idx;
+         return;
     }
 
-    ++iface->super.rx.srq.available;
+    if (wqe_ctr == srq->free_idx + 1) {
```
Comment (yosefe) [inline src/uct/ib/rc/accel/rc_mlx5.inl:182]: pls add ( ) around srq->free_idx + 1

File: src/uct/ib/rc/accel/rc_mlx5_common.c [hunk 15]
```diff
@@ -62,23 +62,22 @@ ucs_config_field_t uct_rc_mlx5_common_config_table[] = {
...
-   ucs_offsetof(uct_rc_mlx5_iface_common_config_t, linked_list_srq),
-   UCS_CONFIG_TYPE_BOOL},
+  {"SRQ_TOPO", "cyclic,cyclic_emulated",
+   "List of SRQ topology types in order of preference. Supported types are:\n"
+   "\n"
+   "list              SRQ is organized as a buffer containing linked list of WQEs.\n"
+   "\n"
+   "cyclic            SRQ is organized as a continuos array of WQEs. Requires DEVX.\n"
+   "\n"
+   "cyclic_emulated   SRQ is organized as a continuos array of WQEs. Doesn`t require DEVX.",
```
Comment (yosefe) [inline src/uct/ib/rc/accel/rc_mlx5_common.c:72]: SRQ is organized as a continuous array of WQEs, but HW treats it as a linked list. Doesn`t require DEVX.

File: src/uct/ib/rc/accel/rc_mlx5_common.c [hunk 16]
```diff
@@ -62,23 +62,22 @@ ucs_config_field_t uct_rc_mlx5_common_config_table[] = {
...
-  {"LINKED_LIST", "n",
-   "TBD",
-   ucs_offsetof(uct_rc_mlx5_iface_common_config_t, linked_list_srq),
-   UCS_CONFIG_TYPE_BOOL},
+  {"SRQ_TOPO", "cyclic,cyclic_emulated",
+   "List of SRQ topology types in order of preference. Supported types are:\n"
+   "\n"
+   "list              SRQ is organized as a buffer containing linked list of WQEs.\n"
+   "\n"
+   "cyclic            SRQ is organized as a continuos array of WQEs. Requires DEVX.\n"
```
Comment (yosefe) [inline src/uct/ib/rc/accel/rc_mlx5_common.c:70]: continuous

File: src/uct/ib/rc/accel/rc_mlx5_common.c [hunk 17]
```diff
@@ -107,35 +153,48 @@ unsigned uct_rc_mlx5_iface_srq_post_recv(uct_rc_mlx5_iface_common_t *iface)
...
+    uct_ib_mlx5_srq_t *srq   = &iface->rx.srq;
+    uct_rc_iface_t *rc_iface = &iface->super;
+    uct_ib_mlx5_srq_seg_t *seg = NULL;
+    uint16_t count = 0;
+    uint16_t wqe_index, next_index;
+
+    ucs_assert(rc_iface->rx.srq.available > 0);
+
+    wqe_index = srq->ready_idx;
+    seg = uct_ib_mlx5_srq_get_wqe(srq, wqe_index);
```
[the remark this answers is not in the document]
Comment (yosefe) [reply src/uct/ib/rc/accel/rc_mlx5_common.c:179]: unresolved indent

File: src/uct/ib/rc/accel/rc_mlx5_common.c (hunk 3 above)
Comment (yosefe) [reply src/uct/ib/rc/accel/rc_mlx5_common.c:192]: unresolved

File: src/uct/ib/rc/accel/rc_mlx5_common.h [hunk 18]
```diff
@@ -103,6 +103,13 @@
 UCT_RC_MLX5_DECLARE_ATOMIC_LE_HANDLER(32)
 UCT_RC_MLX5_DECLARE_ATOMIC_LE_HANDLER(64)
 
+typedef enum {
```
Comment (yosefe) [inline src/uct/ib/rc/accel/rc_mlx5_common.h:106]: 2 space lines in H file (before/after)

File: src/uct/ib/rc/accel/rc_mlx5_iface.c [hunk 19]
```diff
@@ -356,6 +346,33 @@ static ucs_status_t uct_rc_mlx5_iface_tag_recv_cancel(uct_iface_h tl_iface,
...
+{
+    int i;
+
+    for (i = 0; i < config->srq_topo.count; ++i) {
+        if (!strcasecmp(config->srq_topo.types[i], "list")) {
+            *topo_p = UCT_RC_MLX5_SRQ_TOPO_LIST;
+            return UCS_OK;
+        } else if (!strcasecmp(config->srq_topo.types[i], "cyclic")) {
+            if (!(md->flags & UCT_IB_MLX5_MD_FLAG_DEVX_RC_SRQ)) {
+                continue;
```
Comment (yosefe) [inline src/uct/ib/rc/accel/rc_mlx5_iface.c:363]: comment: real cyclic list requires DevX support
BTW, should we check some HCA capability flag here?

**Comment by yosefe** [comment]
@dmitrygx can you pls take a look as well?

File: src/uct/ib/rc/accel/rc_mlx5_iface.c (hunk 19 above)
Comment (dmitrygx) [reply src/uct/ib/rc/accel/rc_mlx5_iface.c:363]: > BTW, should we check some HCA capability flag here?

@yosefe if I'm not mistaken, there is no such capability flag. in #4831 just checking only RMP capability flag
I see that there is only `ib_cyclic_striding_wq` which indicates that LIST and CYCLIC MP RQs are supported

File: src/uct/ib/dc/dc_mlx5.c [hunk 20]
```diff
@@ -40,7 +40,7 @@ ucs_config_field_t uct_dc_mlx5_iface_config_sub_table[] = {
      ucs_offsetof(uct_dc_mlx5_iface_config_t, super),
      UCS_CONFIG_TYPE_TABLE(uct_rc_iface_common_config_table)},
 
-    {"RC_", "", NULL,
+    {"RC_", "SRQ_TOPO=list", NULL,
```
Comment (dmitrygx) [inline src/uct/ib/dc/dc_mlx5.c:43]: maybe add the comment why we prefer `list` WQEs representation (e.g. performance reason or etc)?

File: src/uct/ib/rc/accel/rc_mlx5_iface.c [hunk 21]
```diff
@@ -132,16 +134,34 @@ uct_rc_mlx5_iface_poll_tx(uct_rc_mlx5_iface_common_t *iface)
     return 1;
 }
 
-unsigned uct_rc_mlx5_iface_progress(void *arg)
+static unsigned uct_rc_mlx5_iface_progress(void *arg, int flags)
```
Comment (brminich) [inline src/uct/ib/rc/accel/rc_mlx5_iface.c:137]: add inline?

**Comment by Artemy-Mellanox** [comment]
bot:pipe:retest

File: test/gtest/uct/ib/test_rc.cc [hunk 22]
```diff
@@ -905,3 +905,100 @@ UCS_TEST_SKIP_COND_P(test_rc_keepalive, pending,
...
+}
+
+UCS_TEST_SKIP_COND_P(test_rc_srq, reorder_cyclic,
+                     !check_caps(UCT_IFACE_FLAG_AM_BCOPY),
+                     "RC_SRQ_TOPO?=cyclic,cyclic_emulated")
+{
+    test_reorder();
+}
+
+UCT_INSTANTIATE_RC_DC_TEST_CASE(test_rc_srq);
```
Comment (yosefe) [inline test/gtest/uct/ib/test_rc.cc:1003]: space line after

File: test/gtest/uct/ib/test_rc.cc [hunk 23]
```diff
@@ -905,3 +905,100 @@ UCS_TEST_SKIP_COND_P(test_rc_keepalive, pending,
...
+    test_rc_srq() {
+        m_buf8b      = NULL;
+        m_buf8k      = NULL;
+    }
+
+    void init() {
+        test_rc::init();
+
+        m_buf8b = new mapped_buffer(8, 0x1, *m_e1);
+        m_buf8k = new mapped_buffer(8 * 1024, 0x2, *m_e1);
```
Comment (yosefe) [inline test/gtest/uct/ib/test_rc.cc:922]: UCS_KBYTE

File: test/gtest/uct/ib/test_rc.cc [hunk 24]
```diff
@@ -905,3 +905,100 @@ UCS_TEST_SKIP_COND_P(test_rc_keepalive, pending,
...
 
 UCT_INSTANTIATE_RC_TEST_CASE(test_rc_keepalive)
+
+
+#ifdef HAVE_MLX5_HW
+class test_rc_srq : public test_rc {
+public:
+    test_rc_srq() {
+        m_buf8b      = NULL;
+        m_buf8k      = NULL;
```
Comment (yosefe) [inline test/gtest/uct/ib/test_rc.cc:915]: `: m_buf8b(NULL)...`

File: test/gtest/uct/ib/test_rc.cc [hunk 25]
```diff
@@ -905,3 +905,100 @@ UCS_TEST_SKIP_COND_P(test_rc_keepalive, pending,
...
+        status = uct_ep_am_bcopy(m_e1->ep(ep), 0, mapped_buffer::pack, buf, 0);
+        if (status == UCS_ERR_NO_RESOURCE) {
+            short_progress_loop();
+        } else if (status < 0) {
+            ASSERT_UCS_OK((ucs_status_t)status);
+        }
+    }
+
+    void check() {
+        if (has_transport("rc_verbs")) {
```
Comment (yosefe) [inline test/gtest/uct/ib/test_rc.cc:946]: better check for mlx5?

File: test/gtest/uct/ib/test_rc.cc [hunk 26]
```diff
@@ -905,3 +905,100 @@ UCS_TEST_SKIP_COND_P(test_rc_keepalive, pending,
...
+            return;
+        }
+
+        uct_rc_mlx5_iface_common_t *iface;
+        uct_ib_mlx5_srq_seg_t *seg;
+        uct_ib_mlx5_srq_t *srq;
+        unsigned count = 0;
+
```
Comment (yosefe) [inline test/gtest/uct/ib/test_rc.cc:954]: remove one space line

File: test/gtest/uct/ib/test_rc.cc [hunk 27]
```diff
@@ -905,3 +905,100 @@ UCS_TEST_SKIP_COND_P(test_rc_keepalive, pending,
...
+
+        iface = ucs_derived_of(m_e2->iface(), uct_rc_mlx5_iface_common_t);
+        srq = &iface->rx.srq;
+        seg = (uct_ib_mlx5_srq_seg_t *)srq->buf;
```
Comment (yosefe) [inline test/gtest/uct/ib/test_rc.cc:958]: minor: unite declaration and initialization of those vars

File: test/gtest/uct/ib/test_rc.cc [hunk 28]
```diff
@@ -905,3 +905,100 @@ UCS_TEST_SKIP_COND_P(test_rc_keepalive, pending,
...
+
+        do {
```
Comment (yosefe) [inline test/gtest/uct/ib/test_rc.cc:960]: pls comment what it's doing

File: test/gtest/uct/ib/test_rc.cc [hunk 29]
```diff
@@ -905,3 +905,100 @@ UCS_TEST_SKIP_COND_P(test_rc_keepalive, pending,
...
+            send(1, m_buf8b);
+        }
+
+        flush();
+        check();
+    }
+
+    void cleanup() {
+        delete(m_buf8b);
+        delete(m_buf8k);
```
Comment (yosefe) [inline test/gtest/uct/ib/test_rc.cc:981]: can remove ( )

File: src/uct/ib/rc/accel/rc_mlx5_common.c (hunk 3 above)
Comment (yosefe) [reply src/uct/ib/rc/accel/rc_mlx5_common.c:193]: unresolved: move it outside the loop?

File: src/uct/ib/rc/accel/rc_mlx5_common.c (hunk 17 above)
[the remark this answers is not in the document]
Comment (yosefe) [reply src/uct/ib/rc/accel/rc_mlx5_common.c]: unresolved indent

File: src/uct/ib/rc/accel/rc_mlx5_common.c (hunk 3 above)
Comment (Artemy-Mellanox) [reply src/uct/ib/rc/accel/rc_mlx5_common.c:193]: will yield value next to one we need

Comment (yosefe) [reply src/uct/ib/rc/accel/rc_mlx5_common.c:193]: is it possible to compute wqe_index outside the loop, from the value of next_index ?

Comment (Artemy-Mellanox) [reply src/uct/ib/rc/accel/rc_mlx5_common.c:193]: no. can try rewrite loop logic:
```
@@ -180,17 +180,17 @@ unsigned uct_rc_mlx5_iface_srq_post_recv_ll(uct_rc_mlx5_iface_common_t *iface)
     seg       = uct_ib_mlx5_srq_get_wqe(srq, wqe_index);

     for (;;) {
-        next_index = ntohs(seg->srq.next_wqe_index);
-        if (next_index == (srq->free_idx & srq->mask)) {
+        if (ntohs(seg->srq.next_wqe_index) == (srq->free_idx & srq->mask)) {
             break;
         }
-        seg = uct_ib_mlx5_srq_get_wqe(srq, next_index);
+
+        wqe_index = ntohs(seg->srq.next_wqe_index);
+        seg       = uct_ib_mlx5_srq_get_wqe(srq, wqe_index);

         if (uct_rc_mlx5_iface_srq_set_seg(iface, seg) != UCS_OK) {
             break;
         }

-        wqe_index = next_index;
         count++;
     }
```

Comment (yosefe) [reply src/uct/ib/rc/accel/rc_mlx5_common.c:193]: nvm

**Review by yosefe — APPROVED** [review]

**Comment by yosefe** [comment]
@Artemy-Mellanox pls squash

**Review by yosefe — APPROVED** [review]

**Pull request merged**